drivers/SmartThings/zigbee-fan/src/fan-light/init.lua

```diff
@@ -0,0 +1,161 @@
+}
+
+local function can_handle_itm_fanlight(opts, driver, device)
+  for _, fingerprint in ipairs(FINGERPRINTS) do
+    if device:get_manufacturer() == fingerprint.mfr and device:get_model() == fingerprint.model then
+      return true
+    end
+  end
+  return false
+end
+
+local levels_for_speed = {
+  [0] = 0,
+  [1] = 25,
+  [2] = 50,
+  [3] = 100,
+}
+
+local function level_to_speed(level)
+  local speed = level
+  if speed ~= nil then
+    if level == 0 then
+      speed  = 0
+    elseif level  > 0 and level <= 25 then
+      speed = 1
+    elseif level > 25 and level <= 75 then
+      speed = 2
+    else
+      speed = 3
+    end
+  else
+    speed = 0
+  end
+  return speed
+end
+
+-- CAPABILITY HANDLERS
+
+local function on_handler(driver, device, command)
+  local dev = device
+  if command.component == 'light' then
+    dev:send(OnOff.server.commands.On(dev))
+  else
+    local speed = dev:get_field('LAST_FAN_SPD') or 1
+    dev:send(FanControl.attributes.FanMode:write(dev,speed))
+    dev:send(FanControl.attributes.FanMode:read(dev))
+  end
+  dev:send(FanControl.attributes.FanMode:read(dev))
+end
+
+local function off_handler(driver, device, command)
+  local dev = device
+  if command.component == 'light' then
+    dev:send(OnOff.server.commands.Off(dev))
+  else
+    dev:send(FanControl.attributes.FanMode:write(dev,FanControl.attributes.FanMode.OFF))
+  end
+  dev:send(FanControl.attributes.FanMode:read(dev,FanControl.attributes.FanMode.OFF))
+end
+
+local function switch_level_handler(driver, device, command)
+  local dev = device
+  if command.component == 'light' then
+    local level = math.floor(command.args.level/100.0 * 254)
+    dev:send(Level.server.commands.MoveToLevelWithOnOff(dev, level, command.args.rate or 0xFFFF))
+  else
+    local speed = level_to_speed(command.args.level)
+    dev:send(FanControl.attributes.FanMode:write(dev,speed))
+    dev:send(FanControl.attributes.FanMode:read(dev))
+  end
+end
+
+local function fan_speed_handler(driver, device, command)
+  if command.args.speed <= 3 then
+    device:send(FanControl.attributes.FanMode:write(device, command.args.speed))
+    device:send(FanControl.attributes.FanMode:read(device))
+  else
+    local last_speed = 3
+    device:send(FanControl.attributes.FanMode:write(device, last_speed))
+    device:send(FanControl.attributes.FanMode:read(device))
+  end
+end
+
+-- ZIGBEE HANDLERS
+
+local function zb_fan_control_handler(driver, device, value, zb_rx)
+  device:emit_event(capabilities.fanSpeed.fanSpeed(value.value))
+  local evt = capabilities.switch.switch(value.value > 0 and 'on' or 'off', { visibility = { displayed = false } })
+  device:emit_component_event(device.profile.components.main,evt)
+  device:emit_component_event(device.profile.components.main,capabilities.switchLevel.level(levels_for_speed[value.value], { visibility = { displayed = false } }))
+  if value.value > 0 then
+    device:set_field('LAST_FAN_SPD', value.value, {persist = true})
```

Comment:
this value is available to you already using `device:get_latest_state("main", capabilities.fanSpeed.ID, capabilities.fanSpeed.fanSpeed.NAME)`

Reply:
When I apply the code that was guided to me, if I turn on the fan after turning it off, I can't retrieve the last value (fan speed level) and instead retrieve the off state. Can I still use set field & get_field as before?